nmdc_automation/workflow_automation/models.py

```diff
@@ -0,0 +1,274 @@
+from typing import List, Dict, Any, Optional, Set, Union
+
+from nmdc_schema.nmdc import (
+    FileTypeEnum,
+    NucleotideSequencing,
+    MagsAnalysis,
+    MetagenomeAssembly,
+    MetagenomeAnnotation,
+    MetatranscriptomeAssembly,
+    MetatranscriptomeAnnotation,
+    MetatranscriptomeExpressionAnalysis,
+    PlannedProcess,
+    ReadBasedTaxonomyAnalysis,
+    ReadQcAnalysis,
+    WorkflowExecution
+)
+from nmdc_schema import nmdc
+
+
+def workflow_process_factory(record: Dict[str, Any]) -> PlannedProcess:
+    """
+    Factory function to create a PlannedProcess subclass object from a record.
+    Subclasses are determined by the "type" field in the record, and can be
+    either a WorkflowExecution or DataGeneration object.
+    """
+    process_types = {
+        "nmdc:MagsAnalysis": MagsAnalysis,
+        "nmdc:MetagenomeAnnotation": MetagenomeAnnotation,
+        "nmdc:MetagenomeAssembly": MetagenomeAssembly,
+        "nmdc:MetatranscriptomeAnnotation": MetatranscriptomeAnnotation,
+        "nmdc:MetatranscriptomeAssembly": MetatranscriptomeAssembly,
+        "nmdc:MetatranscriptomeExpressionAnalysis": MetatranscriptomeExpressionAnalysis,
+        "nmdc:NucleotideSequencing": NucleotideSequencing,
+        "nmdc:ReadBasedTaxonomyAnalysis": ReadBasedTaxonomyAnalysis,
+        "nmdc:ReadQcAnalysis": ReadQcAnalysis,
+    }
+    record.pop("_id", None)
+    try:
+        cls = process_types[record["type"]]
+    except KeyError:
+        raise ValueError(f"Invalid workflow execution type: {record['type']}")
+    return cls(**record)
+
+
+class WorkflowProcessNode(object):
+    """
+    Class to represent a workflow processing node. This is a node in a tree
+    structure that represents the tree of data generation and
+    workflow execution objects with their associated data objects.
+    """
+    def __init__(self, record: Dict[str, Any], workflow: "WorkflowConfig"):
+        self.parent = None
+        self.children = []
+        self.data_objects_by_type = {}
+        self.workflow = workflow
+        process = workflow_process_factory(record)
+        self.process = process
+
+    def __hash__(self):
+        return hash((self.id, self.type))
+
+    def __eq__(self, other):
+        return self.id == other.id and self.type == other.type
+
+    def add_data_object(self, data_object):
+        self.data_objects_by_type[data_object.data_object_type] = data_object
+
+    @property
+    def id(self):
+        return self.process.id
+
+    @property
+    def type(self):
+        return self.process.type
+
+    @property
+    def name(self):
+        return self.process.name
+
+    @property
+    def has_input(self):
+        return self.process.has_input
+
+    @property
+    def has_output(self):
+        return self.process.has_output
+
+    @property
+    def git_url(self):
+        """ workflow executions have a git_url field, data generations do not"""
+        default_url = "http://github.com/microbiomedata"
```

Comment:
This is not specific enough, don't use this to pull the actual wdl?

Reply:
We do return the actual url when there is one:
`getattr(self.process, "git_url", default_url)` = Get the `git_url` attribute if it is set, otherwise return the default.

Reply:
`self.process` can hold either a WorkflowExecution or a DataGeneration, which does not have a Git url associated with it

Reply:
I changed this to return `None` if there is no `git_url` set, and all the tests pass.